hazelcast-sql/src/test/java/com/hazelcast/sql/impl/client/SqlResubmissionTest.java

```diff
@@ -0,0 +1,228 @@
+    private static final int INITIAL_CLUSTER_SIZE = 1;
+
+    @Parameterized.Parameter
+    public ClusterFailureTestSupport.SingleFailingInstanceClusterFailure clusterFailure;
+
+    @Parameterized.Parameter(1)
+    public ClientSqlResubmissionMode resubmissionMode;
+
+    private HazelcastInstance client;
+
+    private volatile State state;
+    private volatile boolean done;
+
+    /**
+     * Elapsed time from the test start until the last execution, in millis.
+     */
+    private volatile long lastExecutionTime;
+
+    @Parameterized.Parameters(name = "clusterFailure:{0}, mode:{1}")
+    public static Collection<Object[]> parameters() {
+        List<Object[]> res = new ArrayList<>();
+        List<ClusterFailureTestSupport.SingleFailingInstanceClusterFailure> failures = Arrays.asList(
+                new ClusterFailureTestSupport.NodeReplacementClusterFailure(),
+                new ClusterFailureTestSupport.NodeShutdownClusterFailure(),
+                new ClusterFailureTestSupport.NetworkProblemClusterFailure(),
+                new ClusterFailureTestSupport.NodeTerminationClusterFailure()
+        );
+        for (ClientSqlResubmissionMode mode : ClientSqlResubmissionMode.values()) {
+            for (ClusterFailureTestSupport.SingleFailingInstanceClusterFailure failure : failures) {
+                res.add(new Object[]{failure, mode});
+            }
+        }
+        return res;
+    }
+
+    @Before
+    public void initFailure() {
+        clusterFailure.initialize(INITIAL_CLUSTER_SIZE, smallInstanceConfig());
+        createMap(clusterFailure.getNotFailingInstance(), COMMON_MAP_NAME, COMMON_MAP_SIZE, IntHolder::new,
+                IntHolder.class);
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getSqlConfig().setSqlResubmissionMode(resubmissionMode);
+
+        client = clusterFailure.createClient(clientConfig);
+        state = State.BEFORE_EXECUTE;
+        done = false;
+        lastExecutionTime = 0;
+    }
+
+    @Test
+    public void when_failingSelectBeforeAnyDataIsFetched() throws InterruptedException {
+        SqlStatement statement = new SqlStatement("select * from " + COMMON_MAP_NAME);
+        testStatement(statement, shouldFailBeforeAnyDataIsFetched(resubmissionMode));
+    }
+
+    @Test
+    public void when_failingUpdate() throws InterruptedException {
+        SqlStatement statement = new SqlStatement("update " + COMMON_MAP_NAME + " set field = 1");
+        testStatement(statement, shouldFailNonSelectQuery(resubmissionMode));
+    }
+
+    private void testStatement(SqlStatement statement, boolean shouldFail)
+            throws InterruptedException {
+        Thread failingThread = new Thread(cyclicFailure);
+        failingThread.start();
+        try {
+            if (shouldFail) {
+                assertThrows(HazelcastSqlException.class,
+                        () -> executeInLoop(statement, result -> false));
+                done = true;
```

Comment:
If the previous assert fails, we won't assign `true` to `done`, and then `failintThread.join()` below will get stuck. Can we do `done = true` in the finally section at the end?

Reply:
yes